### DISCUSSION TITLE
Sockets and RDMA Code example in ADIOS2

### DISCUSSION THREAD

**Discussion by MuzafarWani** [ornladios/ADIOS2#4348] [General]
I am using ADIOS2 SST with an application but till now I was reading and writing on the same file system. I want to use ADIOS for two applications running on different machines. Can anyone point out examples of code in C or C++ where I can find how to use ADIOS2 for such a case in C or C++. I am not sure how to configure the IP address and the location of data write into the code.

**Comment by eisenhauer** [comment]
The code doesn't necessarily look different in this situation than it does on a shared filesystem (with some caveats).  On a cluster with a shared filesystem when the SST writer starts up it writes its "contact information" into a .sst file on the filesystem.  When the reader starts it reads this file to get that contact information, they get in touch and get started.  When the reader and writer don't share a filesystem, the easiest way to exchange contact information is for each side to have the engine parameter "RegistrationMethod" set to "screen" on each side.  With this setting, instead of writing the contact information to a file, the writer will print it to stdout, and instead of looking for a file, the reader will prompt for it to be entered via stdin.  Normally this lets you, for example, cut and paste the contact information from one window to another.  

The other question is what is in that "contact information".  Normally it does contain an IP address and port, and possibly RDMA info.  All this is setup by the writer, picking an available port within a certain range and choosing an IP address or hostname as best it can.  Unfortunately, a machine might have have multiple network interface, meaning multiple IP addresses, etc.  You might also want to control the port range used, stop RDMA from being used, etc.  We have a utility, sst_conn_tool that can help sort out what options, environment variables, etc. are necessary.   Docs on that tool are here:  [https://adios2.readthedocs.io/en/latest/ecosystem/utilities.html](https://adios2.readthedocs.io/en/latest/ecosystem/utilities.html)

Please check it out and if you have further questions or issues, we're happy to help.

**Reply by MuzafarWani** [reply]
Thanks. Can you point to an example where either the "RegistrationMethod" set to "screen" or the other method where we can share the contact information using a file. I am trying to make it work for a code where the reader and writer don't share the file system. Below is a snippet of the reader and writer code in C

_**Writer**

    // Set the engine to SST 
    adios2_set_engine(io, "SST");
    // Define a single variable for the combined array
    adios2_variable *var_combined = adios2_define_variable(io, "GlobalArray_combined", adios2_type_float, 1, shape,
                                                           start, count, adios2_constant_dims_true);
    adios2_engine *engine = adios2_open(io, "adios2-global-array-1d-sst", adios2_mode_write);
    adios2_begin_step(engine, adios2_step_mode_append, 20.0f, &err);
    adios2_put(engine, var_combined, combined_array, adios2_mode_deferred);
     adios2_end_step(engine);    

        **Reader** 
     adios2_io *io = adios2_declare_io(adios, "input");
    // Set the engine to SST
    adios2_set_engine(io, "SST");
    adios2_engine *engine = adios2_open(io, streamname, adios2_mode_read);
    adios2_begin_step(engine, adios2_step_mode_read, 40.0, &err);
    adios2_variable *var_combined = adios2_inquire_variable(io, "GlobalArray_combined");
   // Set the selection for reading the combined array
   adios2_set_selection(var_combined, 1, start, count);   
   // Initiate reading data in default/deferred mode: data is available after end_step
    adios2_get(engine, var_combined, combined_array, adios2_mode_deferred)
    adios2_end_step(engine);_

**Reply by eisenhauer** [reply]
I don't see a handy example written in C, but you just need to add: 
`adios2_set_parameter(io, "RegistrationMethod", "screen");`
after you create the IO and before the open() call in both the reader and the writer.  (Alternatively it could be done in a .xml file if you are using one.)

**Reply by MuzafarWani** [reply]
Yes it worked with adios2_set_parameter(io, "RegistrationMethod", "screen");  Thank you 
But do you have any example to point to  with .xml with c++

**Reply by eisenhauer** [reply]
I don't see any .xml files with that in it either, but the XML form would be:
`            <parameter key="RegistrationMethod" value="screen"/>`
For our examples/simulations/gray-scott in the source code, something like this in the xml file:
```
<?xml version="1.0"?>
<adios-config>

    <!--============================================
           Configuration for Gray-Scott and GS Plot
        ============================================-->

    <io name="SimulationOutput">
        <engine type="BP5">
            <!-- SST engine parameters -->
            <parameter key="RegistrationMethod" value="screen"/>
        </engine>
    </io>
</adios-config>
```

**Reply by MuzafarWani** [reply]
Thank you
